oak/server/rust/oak_runtime/src/runtime.rs

```diff
@@ -16,34 +16,68 @@
 
 use std::prelude::v1::*;
 
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
 use std::string::String;
 use std::sync::Arc;
 
-use core::sync::atomic::AtomicBool;
+use core::sync::atomic::{AtomicBool, AtomicUsize};
 use core::sync::atomic::Ordering::SeqCst;
 
-use log::error;
+use log::{debug, error};
 
-use oak_abi::OakStatus;
+use oak_abi::{ChannelReadStatus, OakStatus};
 
 use crate::channel;
-use crate::channel::{ChannelReader, ChannelWriter};
+use crate::message::Message;
+use crate::channel::Channel;
 use crate::node::{load_wasm, NodeConfiguration};
 use crate::platform;
 use crate::proto;
 use crate::proto::application::NodeConfiguration_oneof_config_type;
 
-/// Runtime structure for configuring and running a set of Oak nodes.
+#[derive(PartialEq, Eq, Hash)]
```

Comment:
How about the following (trying to simplify things a bit):

```rust
struct NodeId(usize);
struct Node {
  handles: HashSet<Handle>;
  /* ... */
};

struct ChannelId(usize);
struct Channel {
  /* ... */
}

/// All handles have the same representation.
/// Ideally this would also be already randomized (as per my other comment).
struct Handle(usize);

// Possibly come up with a better name...
enum Object {
  ChannelWriter { channel_id: ChannelId },
  ChannelReader { channel_id: ChannelId },
}

struct Runtime {
  nodes: HashMap<NodeId, Node>,
  channels: HashMap<ChannelId, Channel>,

  objects: HashMap<Handle, Object>,
}
```

Alternatively, we could use [`Weak`](https://doc.rust-lang.org/std/rc/struct.Weak.html) as in

```rust
enum Object {
  ChannelWriter { channel: Weak<Channel> }
  ChannelReader{ channel: Weak<Channel> }
}
```

and avoid the intermediate `ChannelId` (basically use the memory address as the id), but I feel that having explicit, printable ids for everything (including nodes and channels) would be useful anyways.

Thoughts?

Also would like to hear @daviddrysdale 's opinion.

Reply:
FTR

I realised that effectively there are already cases in which just having a node talk to its direct channel(reader)s is not sufficient. For instance:

https://github.com/project-oak/oak/blob/ca6eb8608e826640957540f3984fe3fa5c3008c4/oak/server/rust/oak_runtime/src/node/logger.rs#L35

in which in order to wait on channels we need to go back to the Runtime.

That line may as well read

```rust
runtime.wait_on_channels([handle])
```

and shows how the API I am suggesting may look like if we extended to other channel operations, e.g.

```rust
runtime.read(handle)
```